Java Lexer based syntax highlighting

AntonPieper opened the pull request:
I have ported the [C Version](https://github.com/AntonPieper/ShaderEditor/tree/c-highlight) of my Lexer to Java, so that the codebase is in java only. I also simplified the rendering back to an `EditText` with spans. The benefit is that emojis and ligatures are handled correctly. The tokenization by itself is fast enough to be run on every edit. The rendering was slow because of the repeated construction of the spans. To improve this, only the spans around the edited text will be regenerated in this version.

AntonPieper commented:
This solution could then be used to add a second asynchronous semantic analysis step. In this step, a stack could be used to store variables and their types to i.e. highlight local variables, const variables and function invocations (differentiate between struct constructor and normal functions).

markusfisch commented:
This looks amazing! Sorry for not answering - I'm on vacation right now, but will check this as soon as I'm back 😉 

markusfisch commented:
Sorry for the long delay - I finally got around to check all this out. This is quite a handful 😉 

While profiling this branch against the master, I noticed that input was lagging really hard with big amounts of text, which was not due to the lexer, which performed almost always much better than its RegEx counterpart, but due to ligatures, and the custom fonts 😬

So I disabled ligatures by default, and also made the monospace typeface of the system the default font for [Version 2.31.1](https://github.com/markusfisch/ShaderEditor/releases/tag/2.31.1).

But curiously, this change isn't enough for this branch, and I haven't yet found why 😳 
Here's a video of me typing with some nasty big performance hungry shader:

https://github.com/markusfisch/ShaderEditor/assets/485246/259c0df1-287b-41ce-a98c-38d9afada5e9

And here's me doing the same on the master branch:

https://github.com/markusfisch/ShaderEditor/assets/485246/40ac993c-800b-445d-854b-76b2bbcb3d5c

As you can see there's quite some difference 😬 Both use the same font and ligatures are disabled (although the line numbers are missing in the java-highlight branch somehow 🤔). I even took the time for `ShaderEditor.highlight()` and java-highlight was much faster (between 2 and 4 ms, while the RegEx branch took around 80 ms (!)), yet it felt much slower.

Can you confirm this finding on your device with a big shader?

AntonPieper commented:
Hey, thanks for the feedback, I will certainly look into this tomorrow. I did come across a noticable lag on my machine as well. However for me I felt it was dramatically reduced when starting the app with the "profile with low overhead" option. Is this already that version that you are comparing with?

markusfisch commented:
No, I did run both branches without a debugger attached.

AntonPieper commented on app/src/main/java/de/markusfisch/android/shadereditor/widget/ShaderEditor.java:
```diff
@@ -300,13 +311,13 @@ public void afterTextChanged(Editable e) {
 				}
 
 				dirty = true;
+				highlightWithoutChange(e);
```
I haven't tested it yet, but I have a guess. I have changed the highlighting to be instantaneous, because the Lexer can work in realtime. Adding spans does take some time, that's why I added the `diff` method. I think this could be the root of the performance regression

markusfisch replied:
Hm, yes, that may play a role indeed 🤔

AntonPieper commented:
I changed the code to remove all modifications except the lexer-based highlighting, more color groups and error highlighting. I modified the code to only add a span when the color does not match the editor text color (before, I had a slightly off-white color for identifiers). Can you retry the tests on your device and see if that does fix the issue? With these changes, a [big shader](https://www.shadertoy.com/view/Xds3zN) from shadertoy performed perceptually equal to the master branch on my phone (Pixel 7 Pro).

markusfisch commented:
I finally found the time to run the checks and subsequently merge this. It's now working very well and is indeed a lot faster than the RegEx based highlighter. Thanks a lot for all your work on this! :+1:

And sorry that it took so long to do the merge.
